Sources/SwiftSyntax/BumpPtrAllocator.swift

```diff
@@ -0,0 +1,113 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Bump-pointer allocation.
+class BumpPtrAllocator {
+  typealias Slab = UnsafeMutableRawBufferPointer
+
+  static private var SLAB_SIZE: Int = 4096
+  static private var GLOWTH_DELAY: Int = 128
+  static private var SLAB_ALIGNMENT: Int = 8
+
+  private var slabs: [Slab]
+  private var currentPtr: UnsafeMutableRawPointer?
+  private var endPtr: UnsafeMutableRawPointer?
+  private var customSizeSlabs: [Slab]
+  private(set) var totalSizeAllocated: Int
+
+  init() {
+    slabs = []
+    currentPtr = nil
+    endPtr = nil
+    customSizeSlabs = []
+    totalSizeAllocated = 0
+  }
+
+  deinit {
+    /// Deallocate all memory.
+    totalSizeAllocated = 0
+    currentPtr = nil
+    endPtr = nil
+    while !slabs.isEmpty {
+      slabs.removeLast().deallocate()
+    }
+    while !customSizeSlabs.isEmpty {
+      customSizeSlabs.removeLast().deallocate()
+    }
+  }
+
+  /// Calculate the size of the slab at the index.
+  private static func slabSize(at index: Int) -> Int {
+    // Double the slab size every 'GLOWTH_DELAY' slabs.
+    return SLAB_SIZE * (1 << min(30, index / GLOWTH_DELAY))
+  }
+
+  private func startNewSlab() {
+    let newSlabSize = Self.slabSize(at: slabs.count)
+    let newSlab = Slab.allocate(
+      byteCount: newSlabSize, alignment: Self.SLAB_ALIGNMENT)
+    slabs.append(newSlab)
+    currentPtr = newSlab.baseAddress!
+    endPtr = currentPtr!.advanced(by: newSlabSize)
+  }
+
+  /// Allocate 'byteCount' of memory.
+  func allocate(byteCount: Int, alignment: Int) -> UnsafeMutableRawBufferPointer {
+
+    precondition(alignment <= Self.SLAB_ALIGNMENT)
+    guard byteCount > 0 else {
+      return .init(start: nil, count: 0)
+    }
+
+    // Track the total size we allocated.
+    totalSizeAllocated += byteCount
+
+    // Check if the current slab have enough space.
+    if !slabs.isEmpty {
+      let aligned = currentPtr!.alignedUp(toMultipleOf: alignment)
+      if aligned + byteCount <= endPtr! {
+        currentPtr = aligned + byteCount
+        return .init(start: aligned, count: byteCount)
+      }
+    }
+
+    // If the size is too big, allocate a dedicated slab for it.
+    if byteCount >= Self.SLAB_SIZE {
+      let customSlab = Slab.allocate(
+        byteCount: byteCount, alignment: alignment)
+      customSizeSlabs.append(customSlab)
+      return customSlab
+    }
+
+    // Otherwise, start a new slab and try again.
+    startNewSlab()
+    let aligned = currentPtr!.alignedUp(toMultipleOf: alignment)
+    assert(aligned + byteCount <= endPtr!, "Unable to allocate memory!")
+    currentPtr = aligned + byteCount
+    return .init(start: aligned, count: byteCount)
+  }
+
+  /// Allocate a chunk of memory of `MemoryLayout<T>.stride * count'.
+  func allocate<T>(_: T.Type, count: Int) -> UnsafeMutableBufferPointer<T> {
```

Comment:
Maybe I’m wrong but my assumption would be that it’s very common to allocate memory for a single element. Would it make sense to default `count` to 1?

Reply:
If you think it’s better this way, that’s fine. But I imagine that we’ll be passing `1` in 90% of all cases here. But we can always revisit this later if we actually see the call sites.